[Doc] Update sparse compuation doc (#4060)
taichi-dev/taichi: docs/lang/articles/advanced/sparse.md

sidebar_position: 3
---

> # Sparse spatial data structures

Reviewer: Shall we consistently use "Sparse spatial data structures" in this doc when we refer to this term?

Author: Yes, good point.

:::note
Pre-request: please read the [SNodes](lang/articles/misc/internal.md#data-structure-organization) first.
:::

## Motivation

High-resolution 2D/3D grids are often needed in large-scale spatial computation, such as physical simulation, rendering, and 3D reconstruction.

> However, these grids tend to consume a huge amount of memory space and computation if we use dense data structure (see [field](lang/articles/basic/field.md) and [field advanced](lang/articles/advanced/layout.md)).

Reviewer: There are many terms here "grid ... if we use dense data structure", "dense grid", "sparse grid", which might confuse readers. Is the concept "grid" coupled with the data structure or not?

Author: `grid` is a word that is common for computer graphics practitioners.  `grid` is the spatial discretization of the simulation world. In this context, there is no specific data structure coupled with `grid`.

While a programmer may allocate large dense grids to store spatial data (especially physical quantities such as a density or velocity field),
oftentimes, they only care about a small fraction of this dense grid since the rest may be empty space (vacuum or air).

<center>

![BVH](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/bvh_small.png)

</center>

> As shown above, the regions of interest in sparse grids may only occupy a small fraction of the whole bounding box.

Reviewer: Shall we add some transition phrases like "for example" to clearly separate it from the context?

and each pointer points to a 2x2 dense block. The `activate` function could automatically activates `block[1,1]` which includes `x[2,3]` and `block[1,2]` which includes `x[2, 4]`. Other pixels of `block[1,1]` (`x[2,2], x[3, 2], x[3, 3]`) and `block[1,2]` (`x[2, 5], x[3, 4], x[3, 5]`) are also implicitly activated if `pixel` is a dense node. Because all pixels in the dense block share the same activity value.
As the below figure shown, the active blocks and pixels are green. You can write and read the sparse field like a dense field using indices.

<center>

![Pointer](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/pointer_small.png)

</center>

In fact, the sparse field is a Snode tree shown in the following figure. You could use the struct-for loop to loop over the different levels of the SNode tree like the `print_active` function in `pointer.py`. `for i, j in block` would loop over all active `pointer` Snodes (green grids). `for i, j in pixel` would loop over all active `dense` Snodes (green grids).

<center>

![Pointer Snode Tree](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/pointer_tree_small.png)

</center>

> While a null pointer can effectively represent an empty sub-tree, at the leaf level using 64 bits to represent the activity

Reviewer: This part should be moved under `bitmasked SNode`.

Author: This figure is used to show the pointer SNode tree.

Reviewer: Oh I mean paragraphs starting from L116. GitHub shows lines above it, which might be a little confusing :-)

Author: I see. Actually, this paragraph applies to both `pointer + dense` and `pointer + bitmasked` examples.

Reviewer: Yes but the main purpose is to introduce `bitmasked`. So I suggest moving all of them under `bitmasked` instead of splitting them in two places.

If we can leverage such "spatial sparsity" and focus computation on the regions we care about,
we will significantly save storage and computing power.

:::note
The key to leverage spatial sparsity is to replace *dense* grids with *sparse* grids.
:::

The traditional sparse spatial data stucture are [Quadtrees](https://en.wikipedia.org/wiki/Quadtree) (2D) and
[Octrees](https://en.wikipedia.org/wiki/Octree) (3D). Since dereferencing pointers is relatively costly on modern computer architectures, compared to quadtrees and octrees, it is more performance-friendly to use shallower trees with larger branching factors.
[VDB](https://www.openvdb.org/) and [SPGrid](http://pages.cs.wisc.edu/~sifakis/papers/SPGrid.pdf) are such examples.
In Taichi, programmers can compose data structures similar to VDB and SPGrid with SNodes. The advantage of Taichi sparse spatial data structure are
1. The sparse data could be used like a dense data structure (accessing using indices)
2. The sparse data could be looped in parallel.
3. Automatic memory access optimization.

![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/doc/sparse_grids_3d.jpg)

> Figure: A 3D fluid simulation that uses both particles and grids. Left to right: particles, 1x1x1 voxels, 4x4x4 blocks, 16x16x16 blocks.

Author: I'm not sure if we should leave this figure.

pixel = block.dense(ti.ij, (2,2))
pixel.place(x)

@ti.kernel
def activate():
    x[2,3] = 1.0
    x[2,4] = 2.0

@ti.kernel
def print_active():
    for i, j in block:
        print("Active block", i, j)
    # output: Active block 1 1
    #         Active block 1 2
    for i, j in x:
        print('field x[{}, {}] = {}'.format(i, j, x[i, j]))
    # output: field x[2, 2] = 0.000000
    #         field x[2, 3] = 1.000000
    #         field x[3, 2] = 0.000000
    #         field x[3, 3] = 0.000000
    #         field x[2, 4] = 2.000000
    #         field x[2, 5] = 0.000000
    #         field x[3, 4] = 0.000000
    #         field x[3, 5] = 0.000000
```
The code snippet above creates an 8x8 sparse grid, with the top-level being 4x4 pointer arrays (line 2 of `pointer.py`),
and each pointer points to a 2x2 dense block. The `activate` function could automatically activates `block[1,1]` which includes `x[2,3]` and `block[1,2]` which includes `x[2, 4]`. Other pixels of `block[1,1]` (`x[2,2], x[3, 2], x[3, 3]`) and `block[1,2]` (`x[2, 5], x[3, 4], x[3, 5]`) are also implicitly activated if `pixel` is a dense node. Because all pixels in the dense block share the same activity value.

> As the below figure shows, the active blocks and pixels are green. You can write and read the sparse field like a dense field using indices.

Reviewer: ```suggestion
You can write and read the sparse field like a dense field using indices. The below figure shows the active blocks and pixels in green.
```

<center>

![Pointer](https://raw.githubusercontent.com/FantasyVR/public_files/sparse_computation/taichi/doc/pointer_small.png)

</center>

> 

Reviewer: ```suggestion
Executing the `activate()` function automatically activates `block[1,1]`, which includes `x[2,3]`, and `block[1,2]`, which includes `x[2,4]`. Other pixels of `block[1,1]` (`x[2,2], x[3,2], x[3,3]`) and `block[1,2]` (`x[2,5], x[3,4], x[3,5]`) are also implicitly activated because all pixels in the dense block share the same activity value.

```

> In fact, the sparse field is a Snode tree shown in the following figure. You could use the struct-for loop to loop over the different levels of the SNode tree like the `print_active` function in `pointer.py`. `for i, j in block` would loop over all active `pointer` Snodes (green grids). `for i, j in pixel` would loop over all active `dense` Snodes (green grids).

Reviewer: ```suggestion
In fact, the sparse field is a SNode tree shown in the following figure. You could use the struct-for loop to loop over the different levels of the SNode tree like the `print_active()` function in `pointer.py`. `for i, j in block` would loop over all active `pointer` SNodes. `for i, j in pixel` would loop over all active `dense` SNodes.
```

`Snode -> SNode` should also be applied to the figures.
As they are all green, explicitly stating "green grids" out doesn't seem helpful.

- Use `ti.deactivate_all_snodes()` to deactivate all cells of all SNodes with sparsity.

When deactivation happens, the Taichi runtime automatically recycles and zero-fills memory of the deactivated containers.

:::note
For performance reasons, `ti.activate(snode, index)` only activates `snode[index]`.
The programmer must ensure all ancestor containers of `snode[index]` is already active.
Otherwise, this operation results in undefined behavior.

Similarly, `ti.deactivate` ...
- does **not** recursively deactivate all the descendants of a cell.
- does **not** trigger deactivation of its parent container, even if all the children of the parent container are deactivated.
:::

#### 4. Ancestor index query
- Use `ti.rescale_index(descendant_snode/field, ancestor_snode, index)` to compute the ancestor index given a descendant index.

```python
print(ti.rescale_index(x, block1, ti.Vector([7, 3]))) # output: [1, 0]
print(ti.rescale_index(x, block2, [7, 3]))            # output: [3, 1]
print(ti.rescale_index(x, pixel,  [7, 3]))            # output: [7, 3]
print(ti.rescale_index(block1, block2, [3, 1]))       # output: [1, 0]
```

> As noticed, the output of `ti.rescale_index(x, block1, ti.Vector([7, 3]))` is `[1, 0]`. You could also compute the `block1` index given `pixel` index `[7, 3]` as `[7//2//2, 3//2//2]`. However, doing so couples computation code with the internal configuration of data structures (in this case, the size of `block1` containers). Use `ti.rescale_index` to avoid hard-coding internal information of data structures.

Reviewer: ```suggestion
Regarding line 1, you can also compute the `block1` index given `pixel` index `[7, 3]` as `[7//2//2, 3//2//2]`. However, doing so couples computation code with the internal configuration of data structures (in this case, the size of `block1` containers). By using `ti.rescale_index()`, you can avoid hard-coding internal information of data structures.
```
